core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -202,4 +221,29 @@ public int hashCode() {
         result = 31 * result + (unit != null ? unit.hashCode() : 0);
         return result;
     }
+
+    @Override
+    public int compareTo(ByteSizeValue other) {
+        long unitValue = unit.toBytes(1);
+        long otherUnitValue = other.unit.toBytes(1);
+
+        ByteSizeUnit minUnit = unitValue < otherUnitValue ? unit : other.unit;
+        long thisValue = this.convert(minUnit);
+        long otherValue = other.convert(minUnit);
+
+        if (thisValue == otherValue) {
```

Comment:
Sorry for all the trouble with this! It was a mess before and you are only making it better and this is really just another thing to do to make it better but hashcode and equals being consistent is one of those cardinal rules of java....